# PROJECT CAPSTONE - CLUSTERING THE WORLD CAPITALS

## 1. BACK GROUND

### 1.1. Introduction


Sister cities or twin towns is a known concept in today’s world where two or more cities or countries form a sort of agreement to promote their ties. These cities or countries need not be in the same geographical areas.  The concept is mainly aimed at increasing the friendship and understanding between the cities or countries that ultimately helps in increasing trade and tourism. Further details of this concept can be seen on this link. 
In this report, we try to find sister cities for world capitals. We select several capital cities across different continents and try to form groups of sister cities using a popular machine learning clustering algorithm and by using the location data of these cities. 

### 1.2. Research Problem

Currently, the concept of sister cities is more driven by political or social ties between cities or countries. The objective of this report is to form clusters of capital cities that are closer to each other based on some physical attributes.

### 1.3. Interests

This piece of research can be of interest for many different groups some of which are as follow;
(a)	Tourists or travelers,
(b)	Tourism companies or authorities,
(c)	Traders/Trading companies & 
(d)	International students

In [4]:
# first let's imposrt all necessary packages required for this analyses

import numpy as np 

import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json 

from geopy.geocoders import Nominatim 

import requests
from pandas.io.json import json_normalize 

import matplotlib.cm as cm
import matplotlib.colors as colors


from sklearn.cluster import KMeans
import folium

print('Required libraries imported.')

Required libraries imported.


## 2. DATA

We will need two data sets for this reseach report.  
1. Getet a list of all the world capital cities with country name, city name and longitude/latitude information
2. Using this latitude & longitude information, we will get location or venue data for these capital cities using Foursquare API based on which cluster analyses will be performed

### 2.1. Get the World Capital Cities Data

The data is avaibale on Kaggle (https://www.kaggle.com/nikitagrec/world-capitals-gps). The data has country name, capital name of the country, latitude and longitude of the capital cities, country code and the continent name. I already downloaded the csv file (named: concap.csv) from the given link and saved on my computer's hard drive.

In [113]:
concap_data = pd.read_csv("C://Users//khadmu01//Downloads//world-capitals-gps//concap.csv")
print(concap_data.head())
print(concap_data.shape)

                                CountryName        CapitalName  \
0                                Somaliland           Hargeisa   
1  South Georgia and South Sandwich Islands  King Edward Point   
2       French Southern and Antarctic Lands  Port-aux-Français   
3                                 Palestine          Jerusalem   
4                             Aland Islands          Mariehamn   

   CapitalLatitude  CapitalLongitude CountryCode ContinentName  
0         9.550000         44.050000         NaN        Africa  
1       -54.283333        -36.500000          GS    Antarctica  
2       -49.350000         70.216667          TF    Antarctica  
3        31.766667         35.233333          PS          Asia  
4        60.116667         19.900000          AX        Europe  
(245, 6)


### Plot the world capital cities on the map

In [18]:
world_map = folium.Map(location=[200, -200], zoom_start=1.5)

for lat, lng, cap, cont in zip(concap_data['CapitalLatitude'], concap_data['CapitalLongitude'], concap_data['CapitalName'], concap_data['CountryName']):
    label = '{}, {}'.format(cap, cont)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=1,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(world_map)  
    
world_map

The file has data for 245 capital cities. Given that some of these cities will be very small and some may not even have Foursquare data available, so we will be working on a subset of this data. Let's select first 150 capital cities only given in this data

In [116]:
concap_data2 = concap_data[0:150]
concap_data2.shape

(150, 6)

Let's plot these 150 cities on the map

In [147]:
capital_map = folium.Map(location=[200, -200], zoom_start=1.5)

for lat, lng, cap, cont in zip(concap_data2['CapitalLatitude'], concap_data2['CapitalLongitude'], concap_data2['CapitalName'], concap_data2['CountryName']):
    label = '{}, {}'.format(cap, cont)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=1,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(capital_map)  
    
capital_map

### 2.2. Get Location Data from Foursquare

Since we have selected the capital cities that we will be working on, we now will move to extract their location data from Foursquare API using latitude & longitude information.

In [31]:
CLIENT_ID = '' # removed because of confidentiality
CLIENT_SECRET = '' # removed because of confidentiality
VERSION = '' # Foursquare API version


In [32]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 1000 # define radius


We are using a limit of 100 venuues within the radius of 1000 meter so that we can have reasonable (and if possible closer to each other) number of venues for each city. If we don't apply this condition, then some may cities may have a very large number of venue information while some others may have a very small number. This can impact the clutering results. 

In [20]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [49]:
# function that extracts the location data for all the capital cities in the list

def getNearbyVenues(country, capital, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for con, cap, lat, lng in zip(country, capital, latitudes, longitudes):
        print(cap)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            con,
            cap, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Country',
                    'Capital',
                    'Capital_Latitude',
                    'Capital_Longitude',
                    'Venue',
                    'Venue_Latitude',
                    'Venue_Longitude',
                    'Venue_Category']
    
    return(nearby_venues)

Now, let's extract the location data for all 150 capital cities

In [117]:
worldcapital_venues2 = getNearbyVenues(country = concap_data2['CountryName'], capital=concap_data2['CapitalName'],
                                   latitudes=concap_data2['CapitalLatitude'],
                                   longitudes=concap_data2['CapitalLongitude']
                                  )

Hargeisa
King Edward Point
Port-aux-Français
Jerusalem
Mariehamn
Yaren
Marigot
Atafu
El-Aaiún
Kabul
Tirana
Algiers
Pago Pago
Andorra la Vella
Luanda
The Valley
Saint John's
Buenos Aires
Yerevan
Oranjestad
Canberra
Vienna
Baku
Nassau
Manama
Dhaka
Bridgetown
Minsk
Brussels
Belmopan
Porto-Novo
Hamilton
Thimphu
La Paz
Sarajevo
Gaborone
Brasilia
Road Town
Bandar Seri Begawan
Sofia
Ouagadougou
Rangoon
Bujumbura
Phnom Penh
Yaounde
Ottawa
Praia
George Town
Bangui
N'Djamena
Santiago
Beijing
The Settlement
West Island
Bogota
Moroni
Kinshasa
Brazzaville
Avarua
San Jose
Yamoussoukro
Zagreb
Havana
Willemstad
Nicosia
Prague
Copenhagen
Djibouti
Roseau
Santo Domingo
Quito
Cairo
San Salvador
Malabo
Asmara
Tallinn
Addis Ababa
Stanley
Torshavn
Suva
Helsinki
Paris
Papeete
Libreville
Banjul
Tbilisi
Berlin
Accra
Gibraltar
Athens
Nuuk
Saint George's
Hagatna
Guatemala City
Saint Peter Port
Conakry
Bissau
Georgetown
Port-au-Prince
Vatican City
Tegucigalpa
Budapest
Reykjavik
New Delhi
Jakarta
Tehran
Baghdad
Dub

In [118]:
worldcapital_venues2.head()

Country            Capital  \
0                                Somaliland           Hargeisa   
1  South Georgia and South Sandwich Islands  King Edward Point   
2  South Georgia and South Sandwich Islands  King Edward Point   
3  South Georgia and South Sandwich Islands  King Edward Point   
4       French Southern and Antarctic Lands  Port-aux-Français   

   Capital_Latitude  Capital_Longitude                Venue  Venue_Latitude  \
0          9.550000          44.050000        Hiddo - Dhawr        9.551411   
1        -54.283333         -36.500000        جزيرة سندويشة      -54.282935   
2        -54.283333         -36.500000       Bilinmeyen Yer      -54.281560   
3        -54.283333         -36.500000  Tang Ke Lek Harbour      -54.280980   
4        -49.350000          70.216667            Book Time      -49.352470   

   Venue_Longitude           Venue_Category  
0        44.047806  Comfort Food Restaurant  
1       -36.495176                 Beer Bar  
2       -36.506960                Racetrack  
3       -36.508610          Harbor / Marina  
4        70.218711                Bookstore

In [119]:
worldcapital_venues2.shape

(5913, 8)

Let's check how many venues are available for each city

In [125]:
worldcapital_venues2.groupby(['Capital']).count()


Country  Capital_Latitude  Capital_Longitude  Venue  \
Capital                                                                    
Accra                      3                 3                  3      3   
Addis Ababa                3                 3                  3      3   
Algiers                    8                 8                  8      8   
Amman                    100               100                100    100   
Andorra la Vella          47                47                 47     47   
Antananarivo              11                11                 11     11   
Asmara                     9                 9                  9      9   
Astana                    81                81                 81     81   
Athens                   100               100                100    100   
Avarua                     4                 4                  4      4   
Baghdad                    6                 6                  6      6   
Baku                      97                97                 97     97   
Bamako                     4                 4                  4      4   
Bandar Seri Begawan       36                36                 36     36   
Bangui                     8                 8                  8      8   
Banjul                     4                 4                  4      4   
Beijing                   30                30                 30     30   
Beirut                    19                19                 19     19   
Belmopan                   9                 9                  9      9   
Berlin                   100               100                100    100   
Bishkek                   82                82                 82     82   
Bissau                     4                 4                  4      4   
Bogota                    45                45                 45     45   
Brasilia                  42                42                 42     42   
Brazzaville                1                 1                  1      1   
Bridgetown                41                41                 41     41   
Brussels                 100               100                100    100   
Budapest                 100               100                100    100   
Buenos Aires               7                 7                  7      7   
Bujumbura                  2                 2                  2      2   
Cairo                     61                61                 61     61   
Canberra                  30                30                 30     30   
Chisinau                  55                55                 55     55   
Conakry                    6                 6                  6      6   
Copenhagen               100               100                100    100   
Dhaka                      4                 4                  4      4   
Djibouti                   3                 3                  3      3   
Douglas                   60                60                 60     60   
Dublin                    27                27                 27     27   
El-Aaiún                   4                 4                  4      4   
Gaborone                   4                 4                  4      4   
George Town               44                44                 44     44   
Georgetown                 4                 4                  4      4   
Gibraltar                 37                37                 37     37   
Guatemala City            67                67                 67     67   
Hagatna                    4                 4                  4      4   
Hamilton                  14                14                 14     14   
Hargeisa                   1                 1                  1      1   
Havana                     4                 4                  4      4   
Helsinki                 100               100                100    100   
Jakarta                   99                99                 99    

As shown in above table, there are many cities with very few venues available. For example, Accra and Addid Ababa both have only 3 venues available while Amman and Athen have 100 venues. This can create a noise in the data and ultimately it can impact the results when we run the k-means algorithm for clustering the capital cities. 
<br>
To address this issue, we select only those cities for our clustering analyses that have more than 60 venues available. This is to ensure a good base for running clustering algorithm.

In [133]:
worldcapital_venues = worldcapital_venues2[worldcapital_venues2.groupby('Capital')['Capital'].transform('count')>60]
worldcapital_venues.head()

Country    Capital  Capital_Latitude  Capital_Longitude  \
5  Palestine  Jerusalem         31.766667          35.233333   
6  Palestine  Jerusalem         31.766667          35.233333   
7  Palestine  Jerusalem         31.766667          35.233333   
8  Palestine  Jerusalem         31.766667          35.233333   
9  Palestine  Jerusalem         31.766667          35.233333   

                                     Venue  Venue_Latitude  Venue_Longitude  \
5                           Mt. Zion Hotel       31.768550        35.226596   
6                     Mount Zion (הר ציון)       31.772212        35.229783   
7                            City of David       31.774049        35.235800   
8                               yes PLANET       31.762070        35.226095   
9  Jerusalem Cinematheque (סינמטק ירושלים)       31.770279        35.226014   

        Venue_Category  
5                Hotel  
6             Mountain  
7        Historic Site  
8        Movie Theater  
9  Indie Movie Theater

Just to double check that all the selected cities now have more than 60 venues

In [134]:
worldcapital_venues.groupby(['Capital']).count()

Country  Capital_Latitude  Capital_Longitude  Venue  \
Capital                                                               
Amman               100               100                100    100   
Astana               81                81                 81     81   
Athens              100               100                100    100   
Baku                 97                97                 97     97   
Berlin              100               100                100    100   
Bishkek              82                82                 82     82   
Brussels            100               100                100    100   
Budapest            100               100                100    100   
Cairo                61                61                 61     61   
Copenhagen          100               100                100    100   
Guatemala City       67                67                 67     67   
Helsinki            100               100                100    100   
Jakarta              99                99                 99     99   
Jerusalem            82                82                 82     82   
Kuala Lumpur        100               100                100    100   
Kuwait City         100               100                100    100   
Manama               70                70                 70     70   
Mexico City         100               100                100    100   
Minsk               100               100                100    100   
Monaco               88                88                 88     88   
Nairobi              74                74                 74     74   
Nassau               63                63                 63     63   
Nicosia             100               100                100    100   
Oranjestad           92                92                 92     92   
Ottawa              100               100                100    100   
Paris               100               100                100    100   
Phnom Penh           70                70                 70     70   
Podgorica           100               100                100    100   
Prague              100               100                100    100   
Pristina             80                80                 80     80   
Reykjavik           100               100                100    100   
Riga                100               100                100    100   
Rome                100               100                100    100   
Saint Helier         62                62                 62     62   
San Jose            100               100                100    100   
Santiago            100               100                100    100   
Santo Domingo        71                71                 71     71   
Sarajevo             76                76                 76     76   
Seoul               100               100                100    100   
Skopje              100               100                100    100   
Sofia               100               100                100    100   
Tallinn              91                91                 91     91   
Tbilisi              61                61                 61     61   
Tehran              100               100                100    100   
Tirana              100               100                100    100   
Tokyo               100               100                100    100   
Ulaanbaatar         100               100                100    100   
Vienna              100               100                100    100   
Vientiane            86                86                 86     86   
Zagreb              100               100                100    100   

                Venue_Latitude  Venue_Longitude  Venue_Category  
Capital                                                          
Amman                      100              100             100  
Astana                      81               81              81  
Athens                     100              100            

In [135]:
#number of unique categories and capital cities
print('There are {} uniques categories.'.format(len(worldcapital_venues['Venue_Category'].unique())))
print('There are {} uniques capitals cities.'.format(len(worldcapital_venues['Capital'].unique())))

There are 398 uniques categories.
There are 50 uniques capitals cities.


Finally, we settle with 50 cities each having between 60 to 100 venues. Total unique category venues are 398.

## 3. CLUSTERING

### 3.1. Prepare data for Clustering

Clustering will be done based on the Venue category data but that is a categorical varibale that cannot beused for clustering as is. We need to transform this categorical variable into numeric or quatitative variable. For this we will use one-hot-encoding techniqu. 

In [136]:
# one hot encoding
worldcapital_onehot = pd.get_dummies(worldcapital_venues[['Venue_Category']], prefix="", prefix_sep="")

# add Capital City column back to dataframe
worldcapital_onehot['Capital'] = worldcapital_venues['Capital']

# move Capital City column to the first column
fixed_columns = [worldcapital_onehot.columns[-1]] + list(worldcapital_onehot.columns[:-1])
worldcapital_onehot = worldcapital_onehot[fixed_columns]

worldcapital_onehot.head()

Capital  Accessories Store  Adult Boutique  African Restaurant  \
5  Jerusalem                  0               0                   0   
6  Jerusalem                  0               0                   0   
7  Jerusalem                  0               0                   0   
8  Jerusalem                  0               0                   0   
9  Jerusalem                  0               0                   0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
5                    0             0       0                 0   
6                    0             0       0                 0   
7                    0             0       0                 0   
8                    0             0       0                 0   
9                    0             0       0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
5                       0            0           0                    0   
6                       0            0           0                    0   
7                       0            0           0                    0   
8                       0            0           0                    0   
9                       0            0           0                    0   

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  \
5                     0                 0                   0   
6                     0                 0                   0   
7                     0                 0                   0   
8                     0                 0                   0   
9                     0                 0                   0   

   Australian Restaurant  Austrian Restaurant  Auvergne Restaurant  BBQ Joint  \
5                      0                    0                    0          0   
6                      0                    0                    0          0   
7                      0                    0                    0          0   
8                      0                    0                    0          0   
9                      0                    0                    0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
5           0           0       0     0    0               0   
6           0           0       0     0    0               0   
7           0           0       0     0    0               0   
8           0           0       0     0    0               0   
9           0           0       0     0    0               0   

   Basketball Court  Bath House  Beach  Beach Bar  Bed & Breakfast  Beer Bar  \
5                 0           0      0          0                0         0   
6                 0           0      0          0                0         0   
7                 0           0      0          0                0         0   
8                 0           0      0          0                0         0   
9                 0           0      0          0                0         0   

   Beer Garden  Beer Store  Belarusian Restaurant  Belgian Restaurant  \
5            0           0                      0                   0   
6            0           0                      0                   0   
7            0           0                      0                   0   
8            0           0                      0                   0   
9            0           0                      0                   0   

   Betting Shop  Big Box Store  Bike Rental / Bike Share  Bike Shop  Bistro  \
5             0              0                         0          0       0   
6             0              0                         0          0       0   
7             0              0                         0          0       0   
8             0              0                         0          0       0   
9             0              0                         0          0       0   

   Board Shop  Boarding House  Boat Rental  Boat or Ferry  Bookstore  \


After one-hot-encoding, next let's group rows by Capital cities by taking the mean of the frequency of occurrence of each category

In [137]:
worldcapital_grouped = worldcapital_onehot.groupby('Capital').mean().reset_index()
worldcapital_grouped

Capital  Accessories Store  Adult Boutique  African Restaurant  \
0            Amman               0.00            0.00            0.000000   
1           Astana               0.00            0.00            0.000000   
2           Athens               0.00            0.00            0.000000   
3             Baku               0.00            0.00            0.000000   
4           Berlin               0.00            0.00            0.000000   
5          Bishkek               0.00            0.00            0.000000   
6         Brussels               0.00            0.00            0.010000   
7         Budapest               0.00            0.00            0.000000   
8            Cairo               0.00            0.00            0.000000   
9       Copenhagen               0.00            0.00            0.000000   
10  Guatemala City               0.00            0.00            0.000000   
11        Helsinki               0.00            0.00            0.000000   
12         Jakarta               0.00            0.00            0.000000   
13       Jerusalem               0.00            0.00            0.000000   
14    Kuala Lumpur               0.00            0.00            0.000000   
15     Kuwait City               0.00            0.00            0.000000   
16          Manama               0.00            0.00            0.000000   
17     Mexico City               0.00            0.00            0.000000   
18           Minsk               0.00            0.00            0.000000   
19          Monaco               0.00            0.00            0.000000   
20         Nairobi               0.00            0.00            0.067568   
21          Nassau               0.00            0.00            0.000000   
22         Nicosia               0.00            0.00            0.000000   
23      Oranjestad               0.00            0.00            0.000000   
24          Ottawa               0.00            0.01            0.000000   
25           Paris               0.00            0.00            0.000000   
26      Phnom Penh               0.00            0.00            0.000000   
27       Podgorica               0.00            0.00            0.000000   
28          Prague               0.00            0.00            0.000000   
29        Pristina               0.00            0.00            0.000000   
30       Reykjavik               0.00            0.00            0.000000   
31            Riga               0.00            0.00            0.000000   
32            Rome               0.01            0.00            0.000000   
33    Saint Helier               0.00            0.00            0.000000   
34        San Jose               0.00            0.00            0.000000   
35        Santiago               0.00            0.00            0.000000   
36   Santo Domingo               0.00            0.00            0.000000   
37        Sarajevo               0.00            0.00            0.000000   
38           Seoul               0.00            0.00            0.000000   
39          Skopje               0.00            0.00            0.000000   
40           Sofia               0.00            0.00            0.000000   
41         Tallinn               0.00            0.00            0.000000   
42         Tbilisi               0.00            0.00            0.000000   
43          Tehran               0.00            0.00            0.000000   
44          Tirana               0.00            0.00            0.000000   
45           Tokyo               0.00            0.00            0.000000   
46     Ulaanbaatar               0.00            0.00            0.000000   
47          Vienna               0.00            0.00            0.000000   
48       Vientiane               0.00            0.00            0.000000   
49          Zagreb               0.00            0.00            0.000000   

    American Restaurant  Antique Shop    Arcade  Arepa Restaurant  \
0             

In [138]:
worldcapital_grouped.shape

(50, 399)

The data is now ready to run cluster algorithm. However, let's also map city wise most common venues. This will help us later in explaining the cluster results by comparing capital cities in different clusters according to most common venues.

In [139]:
# function to get the most common venues
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [140]:
num_top_venues = 7

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Capital']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
worldcapital_venues_sorted = pd.DataFrame(columns=columns)
worldcapital_venues_sorted['Capital'] = worldcapital_grouped['Capital']

for ind in np.arange(worldcapital_grouped.shape[0]):
    worldcapital_venues_sorted.iloc[ind, 1:] = return_most_common_venues(worldcapital_grouped.iloc[ind, :], num_top_venues)

worldcapital_venues_sorted.head()

Capital 1st Most Common Venue      2nd Most Common Venue  \
0   Amman                  Café  Middle Eastern Restaurant   
1  Astana           Coffee Shop                       Café   
2  Athens                   Bar                       Café   
3    Baku            Restaurant                   Tea Room   
4  Berlin                 Hotel             History Museum   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Italian Restaurant         Historic Site             Bookstore   
1            Restaurant    Italian Restaurant     Electronics Store   
2           Coffee Shop          Dessert Shop               Theater   
3                 Hotel                  Café           Coffee Shop   
4                 Plaza                Museum           Art Gallery   

  6th Most Common Venue        7th Most Common Venue  
0        Breakfast Spot          Arts & Crafts Store  
1           Karaoke Bar                        Diner  
2          Cocktail Bar                    Bookstore  
3    Turkish Restaurant  Eastern European Restaurant  
4            Art Museum                 Concert Hall

### 3.2. Perform k-menas Clustering

In [141]:
# set number of clusters
kclusters = 5

worldcapital_grouped_clustering = worldcapital_grouped.drop('Capital', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(worldcapital_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 2, 3, 2, 4, 1, 3, 3, 0, 3, 2, 3, 3, 4, 3, 0, 2, 4, 3, 2, 1, 3,
       1, 3, 1, 4, 1, 2, 3, 2, 3, 4, 4, 4, 3, 3, 3, 2, 1, 2, 3, 3, 4, 0,
       2, 4, 1, 4, 2, 2])

Now we will add cluster labels into our data frame where we have arranged capital cities with their most common venue

In [142]:
# add clustering labels
worldcapital_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
worldcapital_venues_sorted.head()

Cluster Labels Capital 1st Most Common Venue      2nd Most Common Venue  \
0               0   Amman                  Café  Middle Eastern Restaurant   
1               2  Astana           Coffee Shop                       Café   
2               3  Athens                   Bar                       Café   
3               2    Baku            Restaurant                   Tea Room   
4               4  Berlin                 Hotel             History Museum   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Italian Restaurant         Historic Site             Bookstore   
1            Restaurant    Italian Restaurant     Electronics Store   
2           Coffee Shop          Dessert Shop               Theater   
3                 Hotel                  Café           Coffee Shop   
4                 Plaza                Museum           Art Gallery   

  6th Most Common Venue        7th Most Common Venue  
0        Breakfast Spot          Arts & Crafts Store  
1           Karaoke Bar                        Diner  
2          Cocktail Bar                    Bookstore  
3    Turkish Restaurant  Eastern European Restaurant  
4            Art Museum                 Concert Hall

Nest step is to map these clusters using Folium. But for this we need latitude & longitude information that is not availabe on above data set. However, that is available in the original data set we downloaded from Kaggle. So the above data frame with cluster lable and the most common venue information is merged with the original world capital cities data with latitude & longitude.

In [143]:
# merge worldcapital_venues_sorted data with the original data having latitude & longitude of the capital cities
worldcapital_merged = concap_data2.join(worldcapital_venues_sorted.set_index('Capital'), on='CapitalName')

worldcapital_merged.head() 

CountryName        CapitalName  \
0                                Somaliland           Hargeisa   
1  South Georgia and South Sandwich Islands  King Edward Point   
2       French Southern and Antarctic Lands  Port-aux-Français   
3                                 Palestine          Jerusalem   
4                             Aland Islands          Mariehamn   

   CapitalLatitude  CapitalLongitude CountryCode ContinentName  \
0         9.550000         44.050000         NaN        Africa   
1       -54.283333        -36.500000          GS    Antarctica   
2       -49.350000         70.216667          TF    Antarctica   
3        31.766667         35.233333          PS          Asia   
4        60.116667         19.900000          AX        Europe   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0             NaN                   NaN                   NaN   
1             NaN                   NaN                   NaN   
2             NaN                   NaN                   NaN   
3             4.0         Historic Site                 Hotel   
4             NaN                   NaN                   NaN   

  3rd Most Common Venue     4th Most Common Venue 5th Most Common Venue  \
0                   NaN                       NaN                   NaN   
1                   NaN                       NaN                   NaN   
2                   NaN                       NaN                   NaN   
3            Restaurant  Mediterranean Restaurant                  Park   
4                   NaN                       NaN                   NaN   

  6th Most Common Venue 7th Most Common Venue  
0                   NaN                   NaN  
1                   NaN                   NaN  
2                   NaN                   NaN  
3    Italian Restaurant          Burger Joint  
4                   NaN                   NaN

Since, we did cluster analyses based on 50 cities only that had more that 60 venues (each) so we see may rown with missing values in the above table. So, let's drop all rows with missing values.

In [144]:
# Since, we finally selected 50 cities only that had more that 60 venues (each) so we ave some missing rows as s
worldcapital_merged = worldcapital_merged.dropna(axis=0)
worldcapital_merged['Cluster Labels'] = worldcapital_merged['Cluster Labels'].astype(int)
worldcapital_merged

CountryName     CapitalName  CapitalLatitude  \
3                 Palestine       Jerusalem        31.766667   
10                  Albania          Tirana        41.316667   
19                    Aruba      Oranjestad        12.516667   
21                  Austria          Vienna        48.200000   
22               Azerbaijan            Baku        40.383333   
23                  Bahamas          Nassau        25.083333   
24                  Bahrain          Manama        26.233333   
27                  Belarus           Minsk        53.900000   
28                  Belgium        Brussels        50.833333   
34   Bosnia and Herzegovina        Sarajevo        43.866667   
39                 Bulgaria           Sofia        42.683333   
43                 Cambodia      Phnom Penh        11.550000   
45                   Canada          Ottawa        45.416667   
50                    Chile        Santiago       -33.450000   
59               Costa Rica        San Jose         9.933333   
61                  Croatia          Zagreb        45.800000   
64                   Cyprus         Nicosia        35.166667   
65           Czech Republic          Prague        50.083333   
66                  Denmark      Copenhagen        55.666667   
69       Dominican Republic   Santo Domingo        18.466667   
71                    Egypt           Cairo        30.050000   
75                  Estonia         Tallinn        59.433333   
80                  Finland        Helsinki        60.166667   
81                   France           Paris        48.866667   
85                  Georgia         Tbilisi        41.683333   
86                  Germany          Berlin        52.516667   
89                   Greece          Athens        37.983333   
93                Guatemala  Guatemala City        14.616667   
101                 Hungary        Budapest        47.500000   
102                 Iceland       Reykjavik        64.150000   
104               Indonesia         Jakarta        -6.166667   
105                    Iran          Tehran        35.700000   
109                  Israel       Jerusalem        31.766667   
110                   Italy            Rome        41.900000   
112                   Japan           Tokyo        35.683333   
113                  Jersey    Saint Helier        49.183333   
114                  Jordan           Amman        31.950000   
115              Kazakhstan          Astana        51.166667   
116                   Kenya         Nairobi        -1.283333   
119             South Korea           Seoul        37.550000   
120                  Kosovo        Pristina        42.666667   
121                  Kuwait     Kuwait City        29.366667   
122              Kyrgyzstan         Bishkek        42.866667   
123                    Laos       Vientiane        17.966667   
124                  Latvia            Riga        56.950000   
132               Macedonia          Skopje        42.000000   
135                Malaysia    Kuala Lumpur         3.166667   
142                  Mexico     Mexico City        19.433333   
145                  Monaco          Monaco        43.733333   
146                Mongolia     Ulaanbaatar        47.916667   
147              Montenegro       Podgorica        42.433333   

     CapitalLongitude CountryCode    ContinentName  Cluster Labels  \
3           35.233333          PS             Asia               4   
10          19.816667          AL           Europe               2   
19         -70.033333          AW    North America               3   
21          16.366667          AT           Europe               4   
22          49.866667          AZ           Europe               2   
23         -77.350000          BS    North America               3   
24          50.566667          BH             Asia               2   
27          27.566667          BY           Europe               3   
28           4.333333          BE           Europe       

Let see how many capital cities are there in each cluster

In [150]:
worldcapital_merged.groupby(['Cluster Labels'])['CapitalName'].count()

Cluster Labels
0     4
1     7
2    12
3    17
4    11
Name: CapitalName, dtype: int64

The data is ready to map clusters on world map

In [146]:
# create map
map_clusters = folium.Map(location=[200, -200], zoom_start=2)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(worldcapital_merged['CapitalLatitude'], worldcapital_merged['CapitalLongitude'], worldcapital_merged['CapitalName'], worldcapital_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=1,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 4. RESULTS

### Cluster 1

In [151]:
worldcapital_merged.loc[worldcapital_merged['Cluster Labels'] == 0, worldcapital_merged.columns[[1] + list(range(5, worldcapital_merged.shape[1]))]]

CapitalName ContinentName  Cluster Labels 1st Most Common Venue  \
71         Cairo        Africa               0                  Café   
105       Tehran          Asia               0                  Café   
114        Amman          Asia               0                  Café   
121  Kuwait City          Asia               0                  Café   

         2nd Most Common Venue      3rd Most Common Venue  \
71                       Plaza                Coffee Shop   
105         Persian Restaurant                Coffee Shop   
114  Middle Eastern Restaurant         Italian Restaurant   
121                Coffee Shop  Middle Eastern Restaurant   

    4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
71    Egyptian Restaurant               Theater           Pastry Shop   
105        Sandwich Place               Theater           Art Gallery   
114         Historic Site             Bookstore        Breakfast Spot   
121   Japanese Restaurant    Italian Restaurant            Hookah Bar   

    7th Most Common Venue  
71              Hotel Bar  
105           Pastry Shop  
114   Arts & Crafts Store  
121           Pizza Place

### Cluster 2

In [152]:
worldcapital_merged.loc[worldcapital_merged['Cluster Labels'] == 1, worldcapital_merged.columns[[1] + list(range(5, worldcapital_merged.shape[1]))]]

CapitalName    ContinentName  Cluster Labels 1st Most Common Venue  \
43    Phnom Penh             Asia               1           Coffee Shop   
45        Ottawa  Central America               1           Coffee Shop   
64       Nicosia           Europe               1           Coffee Shop   
116      Nairobi           Africa               1           Coffee Shop   
119        Seoul             Asia               1           Coffee Shop   
122      Bishkek             Asia               1           Coffee Shop   
146  Ulaanbaatar             Asia               1           Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
43     Chinese Restaurant      Asian Restaurant                 Hotel   
45                   Café            Restaurant      Tapas Restaurant   
64       Greek Restaurant                   Bar                  Café   
116                  Café    African Restaurant                 Hotel   
119                  Café     Korean Restaurant    Italian Restaurant   
122                  Café      Asian Restaurant                 Hotel   
146            Restaurant                  Café                Bakery   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  
43    Japanese Restaurant                  Café       Thai Restaurant  
45                  Hotel            Food Truck    Italian Restaurant  
64               Wine Bar    Italian Restaurant            Restaurant  
116                   Bar  Fast Food Restaurant        Ice Cream Shop  
119                  Park        Scenic Lookout     French Restaurant  
122    Turkish Restaurant   Japanese Restaurant                   Bar  
146    Italian Restaurant                   Pub                Lounge

### Cluster 3

In [153]:
worldcapital_merged.loc[worldcapital_merged['Cluster Labels'] == 2, worldcapital_merged.columns[[1] + list(range(5, worldcapital_merged.shape[1]))]]

CapitalName    ContinentName  Cluster Labels 1st Most Common Venue  \
10           Tirana           Europe               2          Cocktail Bar   
22             Baku           Europe               2            Restaurant   
24           Manama             Asia               2                  Café   
34         Sarajevo           Europe               2                  Café   
61           Zagreb           Europe               2                  Café   
93   Guatemala City  Central America               2                  Café   
115          Astana             Asia               2           Coffee Shop   
120        Pristina           Europe               2            Restaurant   
123       Vientiane             Asia               2                 Hotel   
132          Skopje           Europe               2                  Café   
145          Monaco           Europe               2    Italian Restaurant   
147       Podgorica           Europe               2                  Café   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10                   Café    Italian Restaurant                   Bar   
22               Tea Room                 Hotel                  Café   
24                  Hotel        Breakfast Spot    Italian Restaurant   
34             Restaurant                 Hotel    Italian Restaurant   
61             Restaurant                   Bar                Bakery   
93            Pizza Place  Fast Food Restaurant            Restaurant   
115                  Café            Restaurant    Italian Restaurant   
120                   Bar                 Hotel          Dessert Shop   
123                  Café      Asian Restaurant                   Bar   
132                 Hotel                   Bar            Restaurant   
145     French Restaurant            Restaurant                 Hotel   
147                 Hotel                   Bar    Italian Restaurant   

        5th Most Common Venue 6th Most Common Venue  \
10                      Hotel                Lounge   
22                Coffee Shop    Turkish Restaurant   
24                Coffee Shop            Restaurant   
34                     Hostel          Cocktail Bar   
61                      Hotel  Gym / Fitness Center   
93                Coffee Shop            Steakhouse   
115         Electronics Store           Karaoke Bar   
120  Mediterranean Restaurant  Fast Food Restaurant   
123               Coffee Shop           Pizza Place   
132             Historic Site    Italian Restaurant   
145                       Bar          Cocktail Bar   
147               Pizza Place  Fast Food Restaurant   

           7th Most Common Venue  
10                    Restaurant  
22   Eastern European Restaurant  
24                        Lounge  
34                       Theater  
61                     BBQ Joint  
93                  Burger Joint  
115                        Diner  
120                         Café  
123            French Restaurant  
132                    Bookstore  
145                       Garden  
147                   Restaurant

### Cluster 4

In [154]:
worldcapital_merged.loc[worldcapital_merged['Cluster Labels'] == 3, worldcapital_merged.columns[[1] + list(range(5, worldcapital_merged.shape[1]))]]

CapitalName    ContinentName  Cluster Labels    1st Most Common Venue  \
19      Oranjestad    North America               3     Caribbean Restaurant   
23          Nassau    North America               3            Boat or Ferry   
27           Minsk           Europe               3             Cocktail Bar   
28        Brussels           Europe               3           Sandwich Place   
39           Sofia           Europe               3              Coffee Shop   
50        Santiago    South America               3                      Bar   
59        San Jose  Central America               3                      Bar   
65          Prague           Europe               3                     Café   
66      Copenhagen           Europe               3       Italian Restaurant   
69   Santo Domingo    North America               3                    Hotel   
75         Tallinn           Europe               3                     Café   
80        Helsinki           Europe               3  Scandinavian Restaurant   
89          Athens           Europe               3                      Bar   
101       Budapest           Europe               3           Clothing Store   
102      Reykjavik           Europe               3                      Bar   
104        Jakarta             Asia               3       Chinese Restaurant   
135   Kuala Lumpur             Asia               3         Malay Restaurant   

    2nd Most Common Venue          3rd Most Common Venue  \
19                    Bar                  Shopping Mall   
23     Seafood Restaurant           Caribbean Restaurant   
27                    Bar                     Restaurant   
28                  Hotel                            Bar   
39             Restaurant  Vegetarian / Vegan Restaurant   
50            Pizza Place             Chinese Restaurant   
59         Sandwich Place                          Hotel   
65                 Bakery          Vietnamese Restaurant   
66                   Café        Scandinavian Restaurant   
69               Pharmacy                 Ice Cream Shop   
75       Asian Restaurant                           Park   
80                  Hotel               Sushi Restaurant   
89                   Café                    Coffee Shop   
101           Coffee Shop                          Hotel   
102    Seafood Restaurant                          Hotel   
104                 Hotel                   Noodle House   
135                 Hotel               Asian Restaurant   

    4th Most Common Venue  5th Most Common Venue    6th Most Common Venue  \
19                  Hotel         Breakfast Spot              Coffee Shop   
23                    Bar                  Hotel                    Beach   
27                   Café                   Park              Coffee Shop   
28       Greek Restaurant              Brasserie              Coffee Shop   
39                 Bakery           Dessert Shop                     Park   
50      Martial Arts Dojo                 Hostel                     Café   
59            Coffee Shop         Ice Cream Shop                     Café   
65                    Bar   Gym / Fitness Center                    Hotel   
66                 Bakery            Coffee Shop     Gym / Fitness Center   
69              BBQ Joint             Restaurant              Pizza Place   
75      Electronics Store             Restaurant                    Hotel   
80    Japanese Restaurant            Coffee Shop                   Bakery   
89           Dessert Shop                Theater             Cocktail Bar   
101            Restaurant     Chinese Restaurant                Multiplex   
102                  Café            Coffee Shop  Scandinavian Restaurant   
104    Seafood Restaurant  Indonesian Restaurant               Restaurant   
135       Thai Restaurant         Clothing Store       Chinese Restaurant   

             7th Most Common Venue  
19                      Restaurant  
23            Fast Food 

### Cluster 5

In [155]:
worldcapital_merged.loc[worldcapital_merged['Cluster Labels'] == 4, worldcapital_merged.columns[[1] + list(range(5, worldcapital_merged.shape[1]))]]

CapitalName    ContinentName  Cluster Labels 1st Most Common Venue  \
3       Jerusalem             Asia               4         Historic Site   
21         Vienna           Europe               4                  Café   
81          Paris           Europe               4                 Hotel   
85        Tbilisi           Europe               4                 Hotel   
86         Berlin           Europe               4                 Hotel   
109     Jerusalem             Asia               4         Historic Site   
110          Rome           Europe               4    Italian Restaurant   
112         Tokyo             Asia               4         Historic Site   
113  Saint Helier           Europe               4                 Hotel   
124          Riga           Europe               4            Restaurant   
142   Mexico City  Central America               4    Mexican Restaurant   

           2nd Most Common Venue 3rd Most Common Venue  \
3                          Hotel            Restaurant   
21                         Hotel                 Plaza   
81             French Restaurant                 Plaza   
85          Caucasian Restaurant            Restaurant   
86                History Museum                 Plaza   
109                        Hotel            Restaurant   
110                        Plaza        Ice Cream Shop   
112              Soba Restaurant     Convenience Store   
113                  Coffee Shop                   Pub   
124  Eastern European Restaurant                 Hotel   
142                       Museum            Art Museum   

        4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
3    Mediterranean Restaurant                  Park    Italian Restaurant   
21                     Museum      Asian Restaurant                   Bar   
81        Japanese Restaurant         Historic Site               Theater   
85            Bed & Breakfast           Bus Station           Supermarket   
86                     Museum           Art Gallery            Art Museum   
109  Mediterranean Restaurant                  Park    Italian Restaurant   
110       Monument / Landmark        Sandwich Place              Boutique   
112               Coffee Shop                 Hotel      Ramen Restaurant   
113                Restaurant      Department Store     Fish & Chips Shop   
124                       Bar                 Plaza                  Park   
142                     Hotel        Ice Cream Shop           Coffee Shop   

    7th Most Common Venue  
3            Burger Joint  
21           Concert Hall  
81     Italian Restaurant  
85          Metro Station  
86           Concert Hall  
109          Burger Joint  
110                 Hotel  
112   Japanese Restaurant  
113       Harbor / Marina  
124                  Café  
142   Arts & Crafts Store